Guncode Agony 4: The Great Projectile Purge

SmArtKar opened the pull request:
## About The Pull Request
~~Kept you waitin huh!~~
The projectile refactor is finally here, 4 years later. This PR (almost) completely rewrites projectile logic to be more maintainable and performant.

### Key changes:
* Instead of moving by a fixed amount of pixels, potentially skipping tile corners and being performance-heavy, projectiles now use raymarching in order to teleport through tiles and only visually animate themselves. This allows us to do custom per-projectile animations and makes the code much more reliable, sane and maintainable. You (did not) serve us well, pixel_move.
* Speed variable now measures how many tiles (if SSprojectiles has default values) a projectile passes in a tick instead of being a magical Kevinz Unit:tm: coefficient. pixel_speed_multiplier has been retired because it never had a right to exist in the first place. __This means that downstreams will need to set all of their custom projectiles' speed values to ``pixel_speed_multiplier / speed``__ in order to prevent projectiles from inverting their speed.
* Hitscans no longer operate with spartial vectors and instead only store key points in which the projectile impacted something or changed its angle. This should similarly make the code much easier to work with, as well as fixing some visual jank due to incorrect calculations.
* Projectiles only delete themselves the ***next*** tick after impacting something or reaching their maximum range. Doing so allows them to finish their impact animation and hide themselves between ticks via animation chains. This means that projectiles no longer disappear ~a tile before hitting their target, and that we can finally make impact markers be consistent with where the projectile actually landed instead of being entirely random.

<details>

<summary>Here is an example of how this affects our slowest-moving projectile: Magic Missiles.</summary>


Before:

https://github.com/user-attachments/assets/06b3a980-4701-4aeb-aa3e-e21cd056020e

After:

https://github.com/user-attachments/assets/abe8ed5c-4b81-4120-8d2f-cf16ff5be915

</details>


<details>

<summary>And here is a much faster, and currently jankier, disabler SMG.</summary>


Before:

https://github.com/user-attachments/assets/2d84aef1-0c83-44ef-a698-8ec716587348

After:

https://github.com/user-attachments/assets/2e7c1336-f611-404f-b3ff-87433398d238

</details>

### But how will this affect the ~~trout population~~ gameplay?

Beyond improved visuals, smoother movement and a few minor bugfixes, this should not have a major gameplay impact. If something changed its behavior in an unexpected way or started looking odd, please make an issue report.
Projectile impacts should now be consistent with their visual position, so hitting and dodging shots should be slightly easier and more intuitive.

This PR should be testmerged extensively due to the amount of changes it brings and considerable difficulty in reviewing them. Please contact me to ensure its good to merge.

Closes #71822
Closes #78547
Closes #78871
Closes #83901
Closes #87802
Closes #88073

## Why It's Good For The Game

Our core projectile code is an ungodly abomination that nobody except me, Kapu and Potato dared to poke in the past months (potentially longer). It is laggy, overcomplicated and absolutely unmaintaineable - while a lot of decisions made sense 4 years ago when we were attempting to introduce pixel movement, nowadays they are only acting as major roadblocks for any contributor who is attempting to make projectile behavior that differs from normal in any way.

Huge thanks to Kapu and Potato (Lemon) on the discord for providing insights, ideas and advice throughout the past months regarding potential improvements to projectile code, almost all of which made it in.

## Changelog
:cl:
qol: Projectiles now visually impact their targets instead of disappearing about a tile short of it.
fix: Fixed multiple minor issues with projectile behavior
refactor: Completely rewrote almost all of our projectile code - if anything broke or started looking/behaving oddly, make an issue report!
/:cl:

SmArtKar committed:
Initial commit, cleanup - started vector work

SmArtKar committed:
fucker keeps acting up but i got it nailed down... kinda

SmArtKar committed:
works in northeast dir, teleports everywhere else

SmArtKar committed:
uuuh fuck wip for potato i guess

SmArtKar committed:
it works!

SmArtKar committed:
It works, but dirs other than northeast are still a bit busted

SmArtKar committed:
ITS ALIIIIIVE

SmArtKar committed:
fuck it we animate

SmArtKar committed:
uuuh something yeah

SmArtKar committed:
Hitscans + bugfixes = working projectile code

SmArtKar committed:
oh uh yeah

SmArtKar committed:
Expunges final Kevinz Units:tm: from the codebase

SmArtKar committed:
And swaps the name back

SmArtKar committed:
magic missiles work now

SmArtKar committed:
whuh

SmArtKar committed:
fixes wendigo and removes debug print

SmArtKar committed:
Hitscan fix + autodocs

SmArtKar committed:
comment fix

SmArtKar committed:
Lets clean up a bit more

ZephyrTFA commented:
Brother if you think I'm going to review this you are insane. 
TM it and call it a day

optimumtact commented:
How do you determine what the projectile hits if it raymarches through instantly? Or am i misunderstanding what you're doing.

SmArtKar commented:
> How do you determine what the projectile hits if it raymarches through instantly? Or am i misunderstanding what you're doing.

You are, we're still working on a per-tile basis. The major difference is that we know where the projectile lands this tick in a couple of math operations instead of moving pixel by pixel a hundred times per tick (and a few thousand times per tick for hitscans) and check every pixel if we're on a different tile. Don't think I need to explain why this is beneficial.

Fikou commented:
will this make war ops not lag

SmArtKar commented:
> will this make war ops not lag

I am yet to do proper profiling (byond's profiler does not like me or my projectiles) but it seems like total costs are about halved? Projectile's actual performance impact is minuscule though, but they themselves should be more resilient to the server lagging from other sources (all the explosions and destruction)

SmArtKar committed:
dont spawn them when deleting please thanks

SmArtKar commented:
<details>
<summary>Tracy results</summary>

![tracy-profiler_pQB2ckaS1x](https://github.com/user-attachments/assets/dd151cc3-67e4-4075-990f-ab962e0257bd)

![tracy-profiler_yMNA9qPvkf](https://github.com/user-attachments/assets/b476265c-aa89-42f9-9002-f0f7caaf2208)

</details>
Detailed profiling results - process() is the only call that matters because pixel_move was executed, so there is nothing else to compare. Actual improvement is about 10%, but both are minimal so it doesn't matter all that much.

Rhials commented:
Gun-related flaky failure detected from one of this PR's CI runs. #87755.

SmArtKar commented:
> Gun-related flaky failure detected from one of this PR's CI runs. #87755.

No clue about that, this PR doesn't change anything about projectile's damage handling - on_hit code was changed in 3rd agony when I cleaned it up and moved damage logic to living mobs

SmArtKar committed:
Fixes embedding stuffs

SmArtKar committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into guncode_agony_chapter_four_the_purge

SmArtKar committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into guncode_agony_chapter_four_the_purge

Ghommie commented on code/controllers/subsystem/processing/projectiles.dm:
```diff
@@ -3,21 +3,6 @@ PROCESSING_SUBSYSTEM_DEF(projectiles)
-			return ..()
+	// 5 tiles, in case of overrun
+	var/max_pixels_per_tick = ICON_SIZE_ALL * 5
+	var/pixels_per_decisecond = 32
```
This could use a directive if it has anything to do with the world icon size

Ghommie commented on code/modules/reagents/reagent_containers/cups/soda.dm:
```diff
@@ -76,17 +76,17 @@
-		qdel(src)
+	if(!proj.damage || proj.damage_type != BRUTE)
 		return
+	var/obj/item/trash/can/crushed_can = new /obj/item/trash/can(src.loc)
```
```suggestion
	var/obj/item/trash/can/crushed_can = new /obj/item/trash/can(loc)
```

Ghommie commented on code/modules/projectiles/projectile.dm:
```diff
@@ -139,29 +164,37 @@
 	var/atom/homing_target
-	var/homing_turn_speed = 10 //Angle per tick.
-	var/homing_inaccuracy_min = 0 //in pixels for these. offsets are set once when setting target.
+	/// Angles per move segment, distance is based on SSprojectiles.pixels_per_decisecond
+	/// With pixels_per_decisecond set to 16 and homing_turn_speed, the projectile can turn up to 20 pixels per turf passed
```
```suggestion
	/**
	 * Angles per move segment, distance is based on SSprojectiles.pixels_per_decisecond
	 * With pixels_per_decisecond set to 16 and homing_turn_speed, the projectile can turn up to 20 pixels per turf passed
	 */
```

SmArtKar replied:
I don't think that we need to bloat already big chunks of comments with additional lines for this, autodocs works with multiple ///s

Ghommie replied:
Oh ok

Ghommie commented on code/modules/projectiles/projectile.dm:
```diff
@@ -182,86 +215,105 @@
 	var/list/ignored_factions
-
-	///If defined, on hit we create an item of this type then call hitby() on the hit target with this, mainly used for embedding items (bullets) in targets
+	/// Turf that we have registered connect_loc signal - this is done for performance, as we're moving ~a dozen turfs per tick
+	/// and registering and unregistering signal for every single one of them is stupid. Unregistering the signal from the correct turf in case we get moved by smth else is important
```
```suggestion
	/**
	 * Turf that we have registered connect_loc signal - this is done for performance, as we're moving ~a dozen turfs per tick
	 * and registering and unregistering signal for every single one of them is stupid. Unregistering the signal from the correct turf in case we get moved by smth else is important
	 */
```

Ghommie commented on code/modules/projectiles/projectile.dm:
```diff
@@ -182,86 +215,105 @@
+	var/turf/last_tick_turf
+	/// Remaining pixel movement last tick - used for precise range calculations
+	var/pixels_moved_last_tile = 0
+	/// In order to preserve animations, projectiles are only deleted the tick *after* they impact something.
+	/// Same is applied to reaching the range limit
```
```suggestion
	/**
	 * In order to preserve animations, projectiles are only deleted the tick *after* they impact something.
	 * Same is applied to reaching the range limit
	 */
```

Ghommie commented on code/modules/projectiles/projectile.dm:
```diff
@@ -182,86 +215,105 @@
+	STOP_PROCESSING(SSprojectiles, src)
+	firer = null
+	original = null
+	if (movement_vector)
+		QDEL_NULL(movement_vector)
+	if (beam_points)
+		QDEL_LIST(beam_points)
+	if (last_point)
+		QDEL_NULL(last_point)
```
IIRC, you can use the QDEL_NULL and QDEL_LIST macros without checking the provided parameters / variables. Is this some sort of micro-op?

SmArtKar replied:
You're right nor was this intended to be a microop, but it does technically reduce the amount of unnecessary calls?

Ghommie replied:
Mh

Ghommie commented on code/modules/projectiles/projectile.dm:
```diff
@@ -950,29 +1130,50 @@
 	original = target
+
+	// Trim off excess pixel_x/y by converting them into turf offset
+	if (abs(pixel_x) > ICON_SIZE_X / 2)
```
```suggestion
	if (abs(pixel_x) > ICON_SIZE_X * 0.5)
```

Ghommie commented on code/modules/projectiles/projectile.dm:
```diff
@@ -950,29 +1130,50 @@
+
+	// Trim off excess pixel_x/y by converting them into turf offset
+	if (abs(pixel_x) > ICON_SIZE_X / 2)
+		for (var/i = 1 to floor(abs(pixel_x) + ICON_SIZE_X / 2) / ICON_SIZE_X)
```
```suggestion
		for (var/i in 1 to floor(abs(pixel_x) + ICON_SIZE_X * 0.5) / ICON_SIZE_X)
```

Ghommie commented on code/modules/projectiles/projectile.dm:
```diff
@@ -950,29 +1130,50 @@
+			if (!istype(new_loc))
+				break
+			source_loc = new_loc
+		pixel_x = pixel_x % (ICON_SIZE_X / 2)
```
```suggestion
		pixel_x = pixel_x % (ICON_SIZE_X * 0.5)
```

Ghommie commented on code/modules/projectiles/projectile.dm:
```diff
@@ -950,29 +1130,50 @@
+			source_loc = new_loc
+		pixel_x = pixel_x % (ICON_SIZE_X / 2)
+
+	if (abs(pixel_y) > ICON_SIZE_Y / 2)
```
```suggestion
	if (abs(pixel_y) > ICON_SIZE_Y * 0.5)
```

Ghommie commented on code/modules/projectiles/projectile.dm:
```diff
@@ -950,29 +1130,50 @@
+		pixel_x = pixel_x % (ICON_SIZE_X / 2)
+
+	if (abs(pixel_y) > ICON_SIZE_Y / 2)
+		for (var/i = 1 to floor(abs(pixel_y) + ICON_SIZE_Y / 2) / ICON_SIZE_Y)
```
```suggestion
		for (var/i in 1 to floor(abs(pixel_y) + ICON_SIZE_Y * 0.5) / ICON_SIZE_Y)
```

Ghommie commented on code/modules/projectiles/projectile.dm:
```diff
@@ -950,29 +1130,50 @@
+			if (!istype(new_loc))
+				break
+			source_loc = new_loc
+		pixel_y = pixel_y % (ICON_SIZE_X / 2)
```
```suggestion
		pixel_y = pixel_y % (ICON_SIZE_X * 0.5)
```

Ghommie commented on code/modules/projectiles/projectile.dm:
```diff
@@ -751,191 +764,358 @@
-			trajectory.increment(-trajectory_multiplier)
+	while (pixels_to_move > 0 && isturf(loc) && !QDELETED(src) && !deletion_queued)
+		// Because pixel_x/y represents offset and not actual visual position of the projectile, we add 16 pixels to each and cut the excess because projectiles are not meant to be highly offset by default
+		var/pixel_x_actual = abs(pixel_x) == ICON_SIZE_X / 2 ? pixel_x : (pixel_x + ICON_SIZE_X / 2) % ICON_SIZE_X
+		var/pixel_y_actual = abs(pixel_y) == ICON_SIZE_Y / 2 ? pixel_y : (pixel_y + ICON_SIZE_Y / 2) % ICON_SIZE_Y
```
```suggestion
		var/pixel_x_actual = abs(pixel_x) == ICON_SIZE_X * 0.5 ? pixel_x : (pixel_x + ICON_SIZE_X * 0.5) % ICON_SIZE_X
		var/pixel_y_actual = abs(pixel_y) == ICON_SIZE_Y * 0.5 ? pixel_y : (pixel_y + ICON_SIZE_Y * 0.5) % ICON_SIZE_Y
```

SmArtKar committed:
ghom's review

MrMelbert commented on code/controllers/subsystem/processing/projectiles.dm:
```diff
@@ -3,21 +3,6 @@ PROCESSING_SUBSYSTEM_DEF(projectiles)
-			return TRUE
-		else
-			return ..()
+	// 5 tiles, in case of overrun
+	var/max_pixels_per_tick = ICON_SIZE_ALL * 5
+	var/pixels_per_decisecond = ICON_SIZE_ALL
```
These need to be very well documented so we don't repeat history

MrMelbert commented on code/modules/mob/living/simple_animal/hostile/megafauna/bubblegum.dm:
```diff
@@ -297,9 +297,9 @@ Difficulty: Hard
+/mob/living/simple_animal/hostile/megafauna/bubblegum/bullet_act(obj/projectile/proj)
 	if(BUBBLEGUM_IS_ENRAGED)
-		visible_message(span_danger("[src] deflects the projectile; [p_they()] can't be hit with ranged weapons while enraged!"), span_userdanger("You deflect the projectile!"))
+		visible_message(span_danger("[src] deflects the [proj]! [p_they()] can't be hit with ranged weapons while enraged!"), span_userdanger("You deflect the projectile!"))
```
```suggestion
		visible_message(span_danger("[src] deflects the [proj]! [p_They()] can't be hit with ranged weapons while enraged!"), span_userdanger("You deflect the projectile!"))
```

MrMelbert commented on code/modules/projectiles/projectile.dm:
```diff
@@ -15,30 +16,54 @@
-	var/datum/fired_from = null // the thing that the projectile was fired from (gun, turret, spell)
-	var/suppressed = FALSE //Attack message
-	var/yo = null
+	/// Zone at which the projectile was aimed at
+	var/def_zone = ""
```
Nitpicking your choice of words ...
```suggestion
	/// Zone at which the projectile is aimed at
	var/def_zone = ""
```
...Because the def zone changes in impact

MrMelbert commented on code/modules/projectiles/projectile.dm:
```diff
@@ -15,30 +16,54 @@
-	var/yo = null
+	/// Zone at which the projectile was aimed at
+	var/def_zone = ""
+	/// Atom which shot the projectile
+	var/atom/movable/firer = null
```
```suggestion
	/// Atom who shot the projectile (Not the gun, the guy who shot the gun)
	var/atom/movable/firer = null
```

MrMelbert commented on code/modules/projectiles/projectile.dm:
```diff
@@ -15,30 +16,54 @@
+	var/atom/original = null
+	/// Initial target x coordinate offset of the projectile
 	var/xo = null
-	var/atom/original = null // the original target clicked
-	var/turf/starting = null // the projectile's starting turf
+	/// Initial target y coordinate offset of the projectile
+	var/yo = null
```
I would very much recommend going through some of these vars and marking them as `VAR_FINAL` (or protected/private if possible) if they're internal stuff, it would make it so much easier for future work on projectiles (being able to see at a glance which vars you can override and which you should not mess with)

MrMelbert commented on code/modules/projectiles/projectile.dm:
```diff
@@ -182,86 +215,105 @@
+		QDEL_NULL(last_point)
+	return ..()
+
+/obj/projectile/proc/reduce_range()
```
Le proc docs

MrMelbert commented on code/modules/projectiles/projectile.dm:
```diff
@@ -751,191 +764,358 @@
+				delete_distance = distance_to_move - (pixels_moved_last_tile - ICON_SIZE_ALL)
+
+		if (deletion_queued)
+			//pixel_x -= x_shift * ICON_SIZE_X
+			//pixel_y -= y_shift * ICON_SIZE_Y
```
?

SmArtKar replied:
was needed before i swapped it to end_now, forgor

MrMelbert commented on code/modules/projectiles/projectile.dm:
```diff
@@ -751,191 +764,358 @@
+/// Called every time projectile animates its movement, in case child wants to have custom animations.
+/// Returning TRUE cancels normal animation
+/obj/projectile/proc/move_animate(animate_x, animate_y, animate_time = world.tick_lag, deleting = FALSE)
+	return
```
```suggestion
	return FALSE
```

SmArtKar committed:
Melbert's review

SmArtKar committed:
sanity changes, negative axises no longer break when projectile lands on the border perfectly due to rounding issues

SmArtKar committed:
it works but hitscans are 1 tile away visually from where they should be and I don't want to bootleg shift them

WarerBrow commented:
This just doesen't look right. Laser rojectiles are teleporting through and disappearing at the same time. They ARE NOT being drawed properly, I can't even tell anymore if I hit anything at all.

00-Steven commented on code/modules/projectiles/projectile.dm:
```diff
@@ -751,191 +768,360 @@
+ * Main projectile movement cycle.
+ * Normal behavior moves projectiles in a straight line through tiles, but it gets trickier with homing.
+ * Every pixels_per_decisecond we will stop and call process_homing(), which while a bit rough, does not have a significant performance impact
+ * This proc is needs to be very performant, so do not add overridable logic that can be handled in homing or animations to here.
```
```suggestion
 * This proc needs to be very performant, so do not add overridable logic that can be handled in homing or animations here.
```

SmArtKar commented:
> This just doesen't look right. Laser rojectiles are teleporting through and disappearing at the same time. They ARE NOT being drawed properly, I can't even tell anymore if I hit anything at all.
> 
> p.s. I just double-checked it - when I shoot at nothing, projectile looks ok, but when I shoot at someone - it disappears almost instantly (on the first tile) and there's no hitting effect at all.

Live currently has a bugged version due to rounding issues, its not updated yet. Please do not leave bugs on this page but make separate issues.

SmArtKar committed:
fixes reflectors and hitscan impacts

SmArtKar commented:
I've added myself to projectile.dm codeowners in the latest commit because god knows noone else (who is actively reviewing stuff) is willing to dive into all that math

SmArtKar committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into guncode_agony_chapter_four_the_purge

LemonInTheDark commented on code/datums/position_point_vector.dm:
```diff
@@ -1,15 +1,11 @@
-	P.y = a.y + (b.y - a.y) * 0.5
-	P.z = a.z
-	return P
+	return new /datum/point(_z = a.z, _pixel_x = (a.return_px() + b.return_px()) * 0.5, _pixel_y = (a.return_py() + b.return_py()) * 0.5)
```
why * 0.5 and not / 2

SmArtKar replied:
Per our code standards

LemonInTheDark replied:
our standards are stupid

LemonInTheDark commented on code/datums/position_point_vector.dm:
```diff
@@ -89,143 +87,90 @@
 	if(!isnull(tile_z))
 		z = tile_z
+	if(!isnull(p_x))
+		var/x_offset = SIGN(p_x) * FLOOR(abs(p_x) / ICON_SIZE_X, 1)
```
it is concerning that sign breaks behavior, not relevant here but thought, should we update FLOOR to respect sign (or make a new macro to allow differing behavior?)

LemonInTheDark commented on code/datums/position_point_vector.dm:
```diff
@@ -89,143 +87,90 @@
+/datum/point/proc/increment(p_x, p_y)
+	pixel_x += p_x
+	pixel_y += p_y
+	var/x_offset = SIGN(p_x) * FLOOR(abs(p_x) / ICON_SIZE_X, 1)
```
this feels like something that should be more concise

LemonInTheDark reviewed (changes requested):
readover of vector/point code, will see about projectile logic itself tomorrow(TM)

SmArtKar committed:
puts my sign floor thing into a macro

SmArtKar committed:
oopsie daisy

SmArtKar committed:
*angrily shakes fist at the sky*

SmArtKar committed:
don't really need to initialize hitscanning if we're going to get deleted anyways

SmArtKar committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into guncode_agony_chapter_four_the_purge

SmArtKar committed:
per marx's word, smooth impacts for proletariat

SmArtKar committed:
Far smoother, uses anim chains, added some docs but now it disappears a bit before hitting again?

SmArtKar committed:
a testament to kevinz's sins

SmArtKar commented:
PR is fully complete and only requires a final review from Potato/Lemon before being ready to merge. Side effect of latest commit and proper animation code is that projectiles may now take 2 ticks to delete instead of 1, otherwise in ~half the cases they get deleted on tile border instead.

SmArtKar committed:
fuck!!!

SmArtKar committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into guncode_agony_chapter_four_the_purge

Kapu1178 commented on code/__DEFINES/projectiles.dm:
```diff
@@ -81,12 +81,27 @@
 /// e.g. LASER_SHOTS(12, STANDARD_CELL_CHARGE) means 12 shots
 #define LASER_SHOTS(X, MAX_CHARGE) (((100 * MAX_CHARGE) - ((100 * MAX_CHARGE) % X)) / (100 * X)) // I wish I could just use round, but it can't be used in datum members
+
+/// How far do the projectile hits the prone mob
+#define MAX_RANGE_HIT_PRONE_TARGETS 10
```
wut

SmArtKar replied:
dun blame me, that already was a thing - I just moved it to the defines file

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+			y_to_border = movement_vector.pixel_y > 0 ? (ICON_SIZE_Y - pixel_y_actual) / movement_vector.pixel_y : pixel_y_actual / -movement_vector.pixel_y
+
+		// Figure out how much we need to travel to hit the border
+		var/distance_to_border = !isnull(x_to_border) ? (!isnull(y_to_border) ? min(x_to_border, y_to_border) : x_to_border) : (!isnull(y_to_border) ? y_to_border : -1)
```
you're going to hell for this

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+		// Because pixel_x/y represents offset and not actual visual position of the projectile, we add 16 pixels to each and cut the excess because projectiles are not meant to be highly offset by default
+		var/pixel_x_actual = abs(pixel_x) == ICON_SIZE_X * 0.5 ? pixel_x + ICON_SIZE_X * 0.5 : (pixel_x + ICON_SIZE_X * 0.5) % ICON_SIZE_X
+		var/pixel_y_actual = abs(pixel_y) == ICON_SIZE_Y * 0.5 ? pixel_y + ICON_SIZE_Y * 0.5 : (pixel_y + ICON_SIZE_Y * 0.5) % ICON_SIZE_Y
+		// What distances do we need to move to hit the horizontal/vertical turf border
+		var/x_to_border
+		var/y_to_border
+		// If we're moving strictly up/down/left/right then one of these can be 0 and produce div by zero
+		if (movement_vector.pixel_x)
+			x_to_border = movement_vector.pixel_x > 0 ? (ICON_SIZE_X - pixel_x_actual) / movement_vector.pixel_x : pixel_x_actual / -movement_vector.pixel_x
+		if (movement_vector.pixel_y)
+			y_to_border = movement_vector.pixel_y > 0 ? (ICON_SIZE_Y - pixel_y_actual) / movement_vector.pixel_y : pixel_y_actual / -movement_vector.pixel_y
+
+		// Figure out how much we need to travel to hit the border
+		var/distance_to_border = !isnull(x_to_border) ? (!isnull(y_to_border) ? min(x_to_border, y_to_border) : x_to_border) : (!isnull(y_to_border) ? y_to_border : -1)
+		// Something went extremely wrong
+		if (distance_to_border < 0)
```
```suggestion
		var/distance_to_border = INFINITY
		// What distances do we need to move to hit the horizontal/vertical turf border
		var/x_to_border = INFINITY
		var/y_to_border = INFINITY
		// If we're moving strictly up/down/left/right then one of these can be 0 and produce div by zero
		if (movement_vector.pixel_x)
			x_to_border = movement_vector.pixel_x > 0 ? (ICON_SIZE_X - pixel_x_actual) / movement_vector.pixel_x : pixel_x_actual / -movement_vector.pixel_x
			distance_to_border = x_to_border
		if (movement_vector.pixel_y)
			y_to_border = movement_vector.pixel_y > 0 ? (ICON_SIZE_Y - pixel_y_actual) / movement_vector.pixel_y : pixel_y_actual / -movement_vector.pixel_y
			distance_to_border = min(y_to_border, distance_to_border )
		
		// Something went extremely wrong
		if (distance_to_border == INFINITY)
```

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+		var/y_to_border
+		// If we're moving strictly up/down/left/right then one of these can be 0 and produce div by zero
+		if (movement_vector.pixel_x)
+			x_to_border = movement_vector.pixel_x > 0 ? (ICON_SIZE_X - pixel_x_actual) / movement_vector.pixel_x : pixel_x_actual / -movement_vector.pixel_x
```
honestly might be good to split it up into two vars

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+			distance_to_move = SSprojectiles.pixels_per_decisecond
+
+		// Figure out if we move to the next turf and if so, what its positioning relatively to us is
+		var/x_shift = SIGN(movement_vector.pixel_x) * (!isnull(x_to_border) && distance_to_move >= x_to_border)
+		var/y_shift = SIGN(movement_vector.pixel_y) * (!isnull(y_to_border) && distance_to_move >= y_to_border)
```
see previous
```suggestion
		var/x_shift = (distance_to_move >= x_to_border) ? SIGN(movement_vector.pixel_x) : 0
		var/y_shift = (distance_to_move >= y_to_border) ? SIGN(movement_vector.pixel_y) : 0 
		var/moving_turfs = x_shift || y_shift
```

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+		// Figure out if we move to the next turf and if so, what its positioning relatively to us is
+		var/x_shift = SIGN(movement_vector.pixel_x) * (!isnull(x_to_border) && distance_to_move >= x_to_border)
+		var/y_shift = SIGN(movement_vector.pixel_y) * (!isnull(y_to_border) && distance_to_move >= y_to_border)
+		var/turf/new_turf = (x_shift || y_shift) ? locate(x + x_shift, y + y_shift, z) : loc
+		// Calculate where in the turf we will be when we cross the edge.
+		// This is a projectile variable because its also used in hit VFX
+		entry_x = pixel_x + movement_vector.pixel_x * distance_to_move - x_shift * ICON_SIZE_X
+		entry_y = pixel_y + movement_vector.pixel_y * distance_to_move - y_shift * ICON_SIZE_Y
+		var/delete_distance
+
+		if (x_shift || y_shift)
+			// We've hit an invalid turf, end of a z level or smth went wrong
```
new_turf does not need to be in the outer scope
```suggestion
		// This is a projectile variable because its also used in hit VFX
		entry_x = pixel_x + movement_vector.pixel_x * distance_to_move - x_shift * ICON_SIZE_X
		entry_y = pixel_y + movement_vector.pixel_y * distance_to_move - y_shift * ICON_SIZE_Y
		var/delete_distance

		if (moving_turfs)
			// Calculate where in the turf we will be when we cross the edge.
			var/turf/new_turf = locate(x + x_shift, y + y_shift, z)
			// We've hit an invalid turf, end of a z level or smth went wrong
```

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+				delete_distance = distance_to_move - (pixels_moved_last_tile - ICON_SIZE_ALL)
+
+		if (deletion_queued)
+			// We moved to the next turf first, then impacted something
+			// This means that we need to offset our visual position back to the previous turf, then figure out
+			// how much we moved on the next turf (or we didn't move at all in which case we both shifts are 0 anyways)
+			if (loc == new_turf)
+				pixel_x -= x_shift * ICON_SIZE_X
+				pixel_y -= y_shift * ICON_SIZE_Y
```
```suggestion
			// We moved to the next turf first, then impacted something
			// This means that we need to offset our visual position back to the previous turf, then figure out
			// how much we moved on the next turf (or we didn't move at all in which case we both shifts are 0 anyways)
			if (moving_turfs)
				pixel_x -= x_shift * ICON_SIZE_X
				pixel_y -= y_shift * ICON_SIZE_Y
```

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+		var/turf/new_turf = (x_shift || y_shift) ? locate(x + x_shift, y + y_shift, z) : loc
+		// Calculate where in the turf we will be when we cross the edge.
+		// This is a projectile variable because its also used in hit VFX
+		entry_x = pixel_x + movement_vector.pixel_x * distance_to_move - x_shift * ICON_SIZE_X
+		entry_y = pixel_y + movement_vector.pixel_y * distance_to_move - y_shift * ICON_SIZE_Y
```
I am trusting that this works in reverse

LemonInTheDark replied:
unrelated, why set these if we do not actually enter a new turf? odd

SmArtKar replied:
used for impact animations

LemonInTheDark replied:
but we might not be moving here right? why not inside moving_turfs?

SmArtKar replied:
I should probably doc it a bit better, but this is essentially a way for external sources (see set_angle_centered) to modify projectile's flight path upon impact, and impact_x/y relies on it. This would require us to snowflake non-deleting animations to use entry when we're moving turfs and another value when we're not.

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+		// This is a projectile variable because its also used in hit VFX
+		entry_x = pixel_x + movement_vector.pixel_x * distance_to_move - x_shift * ICON_SIZE_X
+		entry_y = pixel_y + movement_vector.pixel_y * distance_to_move - y_shift * ICON_SIZE_Y
+		var/delete_distance
```
give it an explicit default

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+				return
+			// Similarly with range out deletion, need to calculate how many pixels we can actually move before deleting
+			if (deletion_queued)
+				delete_distance = distance_to_move - (pixels_moved_last_tile - ICON_SIZE_ALL)
```
this can lead to adding to distance_to_move which seems... wrong. is that intended? what is the logic here it's not as well documented

SmArtKar replied:
It cannot, reduce_range does ``pixels_moved_last_tile -= ICON_SIZE_ALL`` when it runs

LemonInTheDark replied:
if pixels_moved_last_tile is < 32 then that means we're subtracting by a negative number so we're adding, right?

SmArtKar replied:
FUCK

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+				pixel_x -= x_shift * ICON_SIZE_X
+				pixel_y -= y_shift * ICON_SIZE_Y
+
+			// Similarly to normal animate code, but use lowered deletion distance instead.
```
this comment indicates otherwise which is why I ask (also like, isn't this wrong distance to deletion is always higher?)

SmArtKar replied:
see above, the comment is correct

LemonInTheDark replied:
it's higher for actually hitting a target at LEAST because we're literally adding, right?

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+			return
+
+		// Prevents long-range high-speed projectiles from ruining the server performance by moving 100 tiles per tick when subsystem is set to a high cap
+		if (CHECK_TICK)
+			// If we ran out of time, add whatever distance we're yet to pass to overrun debt to be processed next tick and break the loop
+			overrun += pixels_to_move
+			return
```
@CHECK_TICK sleeps that doesn't seem like what you'd intend here. you want TICK_CHECK

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+			overrun += pixels_to_move
+			return
+
+		if (tile_limit && (x_shift || y_shift))
```
```suggestion
		if (tile_limit && moving_turfs)
```

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+
+	while (isturf(loc) && !QDELETED(src))
+		process_movement(ICON_SIZE_ALL, hitscan = TRUE)
+		if (CHECK_TICK || paused || QDELETED(src))
```
similarly I assume you want TICK_CHECK to avoid sleeping here

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+/// Creates a new keypoint in which the tracer will split
+/obj/projectile/proc/create_hitscan_point(impact = FALSE, tile_center = FALSE, broken_segment = FALSE)
+	var/atom/handle_atom = last_impact_turf || src
+	var/datum/point/new_point = impact ? new /datum/point(handle_atom.x, handle_atom.y, handle_atom.z, impact_x, impact_y) : RETURN_PRECISE_POINT(tile_center ? loc : src)
```
please at least make the inner ternary its own variable this is depressing

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+		create_hitscan_point()
+	. = ..()
+	if(!isturf(loc) || !isturf(target) || !z || QDELETED(src) || deletion_queued)
+		return .
+	if (!hitscan || isnull(movement_vector) || isnull(beam_points) || free_hitscan_forceMove)
+		return .
```
I am raging against the machine and the machine says there's no reason to return . here unless you want to

SmArtKar replied:
our standards are stupid

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+	last_point = new_point
+
+/obj/projectile/forceMove(atom/target)
+	if (hitscan && !isnull(beam_points))
+		create_hitscan_point()
+	. = ..()
+	if(!isturf(loc) || !isturf(target) || !z || QDELETED(src) || deletion_queued)
+		return .
+	if (!hitscan || isnull(movement_vector) || isnull(beam_points) || free_hitscan_forceMove)
+		return .
```
```suggestion
	if (!hitscan || isnull(beam_points))
		return ..()
	create_hitscan_point()
	. = ..()
	if(!isturf(loc) || !isturf(target) || !z || QDELETED(src) || deletion_queued)
		return .
	if (isnull(movement_vector) || free_hitscan_forceMove)
		return .
```

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+	record_hitscan_start(offset = FALSE)
+
+/obj/projectile/proc/generate_hitscan_tracers(impact = TRUE)
+	if (!length(beam_points))
```
why isnull sometimes and !length others?

SmArtKar replied:
because its length can be 0 and some code must be ran while some other cost should not be

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+	var/datum/point/end_point = beam_points[start_point]
+	var/datum/point/midpoint = point_midpoint_points(start_point, end_point)
+	var/obj/effect/projectile/tracer/tracer_effect = new tracer_type(midpoint.return_turf())
+	tracer_effect.apply_vars(angle_between_points(start_point, end_point), midpoint.pixel_x, midpoint.pixel_y, color, pixel_length_between_points(start_point, end_point) / ICON_SIZE_ALL)
```
is there any reason this can't be using explicit arg names and newlines

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
+	var/list/turf/light_line = get_line(start_point.return_turf(), end_point.return_turf())
+	for (var/turf/light_turf as anything in light_line)
+		var/located_effect = FALSE
+		for(var/obj/effect/abstract/projectile_lighting/light_effect in light_turf)
+			if(light_effect.owner == self_ref)
+				located_effect = TRUE
+				break
```
is this really the best way of doing this?

SmArtKar replied:
can't think of anything much better, we'll have to iterate through all of em anyways

LemonInTheDark replied:
depending on how often this is done it might be worth it to use an assoc list (I don't have a good feel for when this proc is called tho)

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -950,29 +1155,50 @@
-/obj/projectile/proc/preparePixelProjectile(atom/target, atom/source, list/modifiers = null, deviation = 0)
+/obj/projectile/proc/aim_projectile(atom/target, atom/source, list/modifiers = null, deviation = 0)
 	if(!(isnull(modifiers) || islist(modifiers)))
 		stack_trace("WARNING: Projectile [type] fired with non-list modifiers, likely was passed click params.")
```
print them?

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -950,29 +1155,50 @@
 	original = target
+
+	// Trim off excess pixel_x/y by converting them into turf offset
+	if (abs(pixel_x) > ICON_SIZE_X * 0.5)
+		for (var/i in 1 to floor(abs(pixel_x) + ICON_SIZE_X * 0.5) / ICON_SIZE_X)
+			var/turf/new_loc = get_step(source_loc, pixel_x > 0 ? EAST : WEST)
+			if (!istype(new_loc))
+				break
+			source_loc = new_loc
+		pixel_x = pixel_x % (ICON_SIZE_X * 0.5)
+
```
the for loop predicate is real verbose but a while doesn't quite fit here cause you'd need a finishing modulus anyway, can't decide what's ebst

SmArtKar replied:
m pretty sure this is the best tbh

LemonInTheDark commented on code/modules/projectiles/projectile.dm:
```diff
@@ -999,14 +1225,14 @@
 	var/angle = 0
-	var/p_x = LAZYACCESS(modifiers, ICON_X) ? text2num(LAZYACCESS(modifiers, ICON_X)) : ICON_SIZE_X / 2 // ICON_(X|Y) are measured from the bottom left corner of the icon.
-	var/p_y = LAZYACCESS(modifiers, ICON_Y) ? text2num(LAZYACCESS(modifiers, ICON_Y)) : ICON_SIZE_Y / 2 // This centers the target if modifiers aren't passed.
+	var/p_x = LAZYACCESS(modifiers, ICON_X) ? text2num(LAZYACCESS(modifiers, ICON_X)) : ICON_SIZE_X * 0.5 // ICON_(X|Y) are measured from the bottom left corner of the icon.
```
just like, for the record, the standards page is stupid. if you mean to divide by 2 and that is CLEARER then DO THAT, it's a brainworms optimization not a style recommendation. goes for the whole pr

SmArtKar replied:
indeed

LemonInTheDark reviewed (changes requested):
Am operating on the assumption you know that pixel_x/y floor values inserted into them.
My gut is our position and visuals should not be tracked in the same var like this but it seems? intentional.

SmArtKar commented on code/modules/projectiles/projectile.dm:
```diff
@@ -289,43 +347,41 @@
 	else
-		hitx = target.pixel_x + rand(-8, 8)
-		hity = target.pixel_y + rand(-8, 8)
+		impact_x = entry_x + movement_vector?.pixel_x * rand(0, ICON_SIZE_X * 0.5)
+		impact_y = entry_y + movement_vector?.pixel_y * rand(0, ICON_SIZE_Y * 0.5)
```
vector* pixel_x/y are cached sine/cosine to avoid math operations, they don't include magnitude

SmArtKar committed:
first review batch

SmArtKar commented on code/modules/projectiles/projectile.dm:
```diff
@@ -674,268 +735,412 @@
-			trajectory.increment(-trajectory_multiplier)
+	while (pixels_to_move > 0 && isturf(loc) && !QDELETED(src) && !deletion_queued)
+		// Because pixel_x/y represents offset and not actual visual position of the projectile, we add 16 pixels to each and cut the excess because projectiles are not meant to be highly offset by default
+		var/pixel_x_actual = abs(pixel_x) == ICON_SIZE_X * 0.5 ? pixel_x + ICON_SIZE_X * 0.5 : (pixel_x + ICON_SIZE_X * 0.5) % ICON_SIZE_X
```
this is incorrect, the check here is to snowflake for 16px offsets aka projectiles on the border of a tile - not for when its above 32 (although that shouldn't be happening either, that just breaks the math)

SmArtKar committed:
makes my bullshit readable AND reliable

SmArtKar committed:
Now they're planecube friendly too!

LemonInTheDark commented on code/game/objects/effects/temporary_visuals/projectiles/projectile_effects.dm:
```diff
@@ -36,26 +36,21 @@
-	SET_PLANE_EXPLICIT(look, plane, new_loc || src)
-	look.pixel_x = p_x
-	look.pixel_y = p_y
+/obj/effect/projectile/proc/apply_vars(angle_override, p_x = 0, p_y = 0, color_override, scaling = 1, new_plane = null, increment = 0)
+	pixel_x = p_x
+	pixel_y = p_y
```
why do we have the new_plane var if it's gonna be discarded?
unrelated, why nuke the MA here?

SmArtKar replied:
no reason to use MAs, new_plane is a leftover

SmArtKar committed:
brainworms are terminal here

SmArtKar committed:
this works
